ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs

```diff
@@ -0,0 +1,364 @@
+
+        result
+
+    let getFunctionParameterAt parameterIndex (fcsType: FSharpType) =
+        let rec getParameter index (fcsType: FSharpType) =
+            if index = 0 then fcsType.GenericArguments[0]
+            else
+                getParameter (index - 1) fcsType.GenericArguments[1]
+
+        getParameter parameterIndex fcsType
+
+module AnnotationUtil =
+
+    let private isFullyAnnotatedNamedTypeUsage (namedTypeUsage: INamedTypeUsage) =
+        isNull namedTypeUsage.ReferenceName ||
+        isNull namedTypeUsage.ReferenceName.TypeArgumentList ||
+        namedTypeUsage.ReferenceName.TypeArgumentList.TypeUsagesEnumerable
+           |> Seq.exists (fun typeUsage -> typeUsage :? IAnonTypeUsage)
+           |> not
+
+    let rec private isFullyAnnotatedFunctionTypeUsage (functionTypeUsage: IFunctionTypeUsage) =
+        isFullyAnnotatedTypeUsage functionTypeUsage.ArgumentTypeUsage &&
+        isFullyAnnotatedTypeUsage functionTypeUsage.ReturnTypeUsage
+
+    and isFullyAnnotatedTypeUsage (typeUsage: ITypeUsage) =
+        match typeUsage with
+        | :? INamedTypeUsage as namedTypeUsage ->
+            isFullyAnnotatedNamedTypeUsage namedTypeUsage
+        | :? IFunctionTypeUsage as functionTypeUsage ->
+            isFullyAnnotatedFunctionTypeUsage functionTypeUsage
+        | :? IConstrainedTypeUsage as constrainedTypeUsage ->
+            isFullyAnnotatedTypeUsage constrainedTypeUsage.TypeUsage
+        | _ -> false
+
+    let isFullyAnnotatedReturnTypeInfo (returnTypeInfo: IReturnTypeInfo) =
+        isNotNull returnTypeInfo.ReturnType &&
+        isFullyAnnotatedTypeUsage returnTypeInfo.ReturnType
+
+    let rec isFullyAnnotatedPattern (pattern: IFSharpPattern) =
+        match pattern.IgnoreInnerParens() with
+        | :? IUnitPat ->
+            true
+        | :? ITypedPat as typedPat ->
+            isFullyAnnotatedPattern typedPat.Pattern
+        | pattern ->
+            pattern |> PatUtil.getTypeUsage |> ValueOption.exists isFullyAnnotatedTypeUsage
+            || pattern |> PatUtil.getReturnTypeInfo |> ValueOption.exists isFullyAnnotatedReturnTypeInfo
+
+    let isFullyAnnotatedBinding (binding: IBinding) =
+        isNotNull binding.ReturnTypeInfo &&
+        binding.ParametersDeclarationsEnumerable
+        |> Seq.forall (fun parameter ->
+            isFullyAnnotatedPattern parameter.Pattern)
+
+    let isFullyAnnotatedMemberDeclaration (memberDeclaration: IMemberDeclaration) =
+        isNotNull memberDeclaration.ReturnTypeInfo &&
+        isFullyAnnotatedReturnTypeInfo memberDeclaration.ReturnTypeInfo &&
+        memberDeclaration.ParametersDeclarations
+        |> Seq.forall (fun parameter ->
+            isFullyAnnotatedPattern parameter.Pattern)
+
+    // let f x ... =
+    // let f<'a> = ...
+    let isFunctionBinding (binding: IBinding) =
+        binding.HasParameters
+        || binding.HeadPattern.GetNextMeaningfulSibling() :? IPostfixTypeParameterDeclarationList
```

Comment:
I'd not expect a generic value to be considered as a function from F# point of view. Why is this implemented like this?

Reply:
Well, from my point of view it behaves exactly like a function. E.g. it let f<'a> = ... behaves like an let f<'a>() = ...

Reply:
It's indeed compiled to a method but it's not considered a function from F# point of view, i.e. it doesn't have a function type.

Reply:
Should I move it to value annotation action?